Choose sparse/dense type when doing tensordot

lueckem opened the pull request:
Enhancement of sparse.tenserdot: choose return type. See #348.
The goal is to let the user decide of which type the output of tensordot is. The options are returntype="sparse", "dense" or "auto" (sparse if both arguments are sparse, otherwise dense).

To date only works for returntype="sparse" in the case that the first array is dense and the second array is sparse. The other cases will be implemented in the future.

I am very thankful for suggestions for improvement, especially concerning perfomance.

hameerabbasi commented:
Can we raise a `NotImplementedError` for unsupported configurations at present?

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -64,7 +64,7 @@ def linear_loc(coords, shape):
 
 
-def tensordot(a, b, axes=2):
+def tensordot(a, b, axes=2, returntype="auto"):
```
I'd change this to `return_type=None`

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -74,6 +74,9 @@ def tensordot(a, b, axes=2):
         The arrays to perform the :code:`tensordot` operation on.
     axes : tuple[Union[int, tuple[int], Union[int, tuple[int]], optional
         The axes to match when performing the sum.
+    returntype : {"auto", "dense", "sparse"}, optional
```
`{None, COO, numpy.ndarray}`

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -1287,6 +1299,42 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
+        out_data = []
+        out_coords = []
+
+        for oidx1 in range(out_shape[0]):
```
I'd perform the accumulation inline rather than relying on the deduplication/coalescing. I believe this is what is slowing down your performance.

lueckem replied:
Thanks for the suggestions! Could you please elaborate on this one?

hameerabbasi replied:
Well, matrix multiplies are a multiply of some elements plus a sum over a given axis, right? What you're doing is adding all the multiplies in, and relying on the deduplication to do the sum. I would also do the sum here as well, and filter out zero entries from the get-go. This would improve performance by a lot.

lueckem replied:
I am not sure what you mean exactly, but the summing seems to be optimal already because for every row of the output (oidx1) we only iterate over the nonzero elements of the sparse array once (didx). How can you improve that? Could you provide some code or formula for your idea?

To the filtering out: I already tried that and it didnt seem to have any effect on the performance in the case that I tested. As the first array is dense, we should assume that it does not have a lot of zero entries, maybe none at all. Thus, the test for zero would almost always be negative. The additional overhead created by this test seems to be of the same order as the one call of prune in the creation of the COO array.

If you prefer, I can switch to the version with tests for zero and without prune, but I have not seen a performance increase in that change.

hameerabbasi replied:
You are adding way more elements than necessary IIUC, one element is added multiple times. If you follow my suggestion in the other comment you'll see that.

lueckem replied:
You are right! I updated the algorithm accordingly and the performance now matches the return_type=None version.
Unfortunately the test fails in Python 3.6. Any idea why?

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -1287,6 +1299,42 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
+            for didx2 in range(len(data2)):
+                oidx2 = coords2[0, didx2]
+                out_coords.append([oidx1, oidx2])
+                out_data.append(array1[oidx1, coords2[1, didx2]] * data2[didx2])
```
As well as filtering zeros here.

lueckem commented:
> Can we raise a `NotImplementedError` for unsupported configurations at present?

I will implement them in the next few days, so maybe we could just keep the PR open until then?

hameerabbasi commented:
> I will implement them in the next few days, so maybe we could just keep the PR open until then?

Of course. Feel free to work on it when you like. I may be quick to respond; you don't have to be. 😄

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -303,12 +307,20 @@ def _dot(a, b):
+            )
+            coords = np.array(coords).T
+            data = np.array(data, dtype=dtype)
+            return COO(coords, data, shape=out_shape, prune=True)
```
In relation to https://github.com/pydata/sparse/pull/349/files#r426539625, don't rely on `prune=True` here.

hameerabbasi replied:
I'd say your tests should pass with `has_duplicates=False, sorted=True`

hameerabbasi commented on sparse/_coo/common.py:
```diff
@@ -64,7 +64,7 @@ def linear_loc(coords, shape):
 
 
-def tensordot(a, b, axes=2):
+def tensordot(a, b, axes=2, return_type=None):
```
```suggestion
def tensordot(a, b, axes=2, *, return_type=None):
```

hameerabbasi replied:
This will make it so that if NumPy adds any additional `kwargs`, we remain compatible in the future.

lueckem commented:
I added all the missing implementations. The performance of the newly added implementations match the old performance for return_type=None. I added some benchmarks for comparison.

hameerabbasi commented:
This is in, thanks so much @lueckem!
